pkg/beholder/auth.go

```diff
@@ -1,27 +1,158 @@
 package beholder
 
 import (
+	"context"
 	"crypto/ed25519"
+	"encoding/binary"
 	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/credentials"
+)
+
+var (
+	// authHeaderKey is the name of the header that the node authenticator will use to send the auth token
+	authHeaderKey = "X-Beholder-Node-Auth-Token"
+	// authHeaderVersion is the version of the auth header format
+	authHeaderVersion1 = "1"
+	authHeaderVersion2 = "2"
 )
 
-// authHeaderKey is the name of the header that the node authenticator will use to send the auth token
-var authHeaderKey = "X-Beholder-Node-Auth-Token"
+const DefaultAuthHeaderTTL = 1 * time.Minute
+
+type AuthHeaderProvider interface {
+	Credentials() credentials.PerRPCCredentials
+}
+
+// authHeaderPerRPCredentials is a PerRPCCredentials implementation that provides the auth headers
+type authHeaderPerRPCredentials struct {
+	privKey                  ed25519.PrivateKey
+	lastUpdated              time.Time
+	headerTTL                time.Duration
+	requireTransportSecurity bool
+	headers                  map[string]string
+	version                  string
+	mu                       sync.Mutex
+}
+
+// AuthHeaderProviderConfig configures AuthHeaderProvider
+type AuthHeaderProviderConfig struct {
+	HeaderTTL                time.Duration
+	Version                  string
+	RequireTransportSecurity bool
+}
+
+func NewAuthHeaderProvider(privKey ed25519.PrivateKey, config *AuthHeaderProviderConfig) AuthHeaderProvider {
```

Comment:
Another option is to use a method for construction:
```suggestion
func (config AuthHeaderProviderConfig) New(privKey ed25519.PrivateKey) AuthHeaderProvider {
```
This simplifies the callers, which don't have to repeat the package or the type name:
```diff
- beholder.NewAuthHeaderProvider(privKey, &beholder.AuthHeaderProviderConfig{HeaderTTL: time.Minute})
+ beholder.AuthHeaderProviderConfig{HeaderTTL: time.Minute}.New(privKey)
```
And you could keep this simple func around too, for the cases passing `nil`:
```suggestion
func NewAuthHeaderProvider(privKey ed25519.PrivateKey) AuthHeaderProvider {
	return AuthHeaderProviderConfig{}.New(privKey)
```